tachyon/math/polynomials/univariate/icicle/icicle_ntt.cc

```diff
@@ -0,0 +1,78 @@
+#include "tachyon/math/polynomials/univariate/icicle/icicle_ntt.h"
+
+#include "third_party/icicle/include/fields/id.h"
+
+#include "tachyon/base/bit_cast.h"
+#include "tachyon/device/gpu/gpu_enums.h"
+#include "tachyon/math/polynomials/univariate/icicle/icicle_ntt_bn254.h"
+
+namespace tachyon::math {
+
+template <>
+bool IcicleNTT<bn254::Fr>::Init(const bn254::Fr& group_gen) {
+#if FIELD_ID != BN254
+#error Only Bn254 is supported
+#endif
+  ::device_context::DeviceContext ctx{stream_, /*device_id=*/0, mem_pool_};
+  math::BigInt<4> group_gen_big_int = group_gen.ToBigInt();
+  gpuError_t error = tachyon_bn254_initialize_domain(
+      reinterpret_cast<const ::bn254::scalar_t&>(group_gen_big_int), ctx,
+      /*fast_twiddles_mode=*/true);
+  if (error != gpuSuccess) return false;
+
+  config_.reset(new ::ntt::NTTConfig<bn254::Fr>{
+      ctx,
+      base::bit_cast<bn254::Fr>(::bn254::scalar_t::one()),
+      /*batch_size=*/1,
+      /*columns_batch=*/false,
+      /*ordering=*/::ntt::Ordering::kNN,
+      /*are_inputs_on_device=*/false,
+      /*are_outputs_on_device=*/false,
+      /*is_async=*/false,
+      /*ntt_algorithm=*/::ntt::NttAlgorithm::Auto,
+  });
+  return true;
+}
+
+template <>
+bool IcicleNTT<bn254::Fr>::Run(::ntt::NttAlgorithm algorithm,
+                               const BigInt& coset, bn254::Fr* inout, int size,
+                               ::ntt::NTTDir dir) const {
+#if FIELD_ID != BN254
+#error Only Bn254 is supported
+#endif
+
+  // NOTE(chokobole): Manual copy is needed even though
+  // |sizeof(::bn254::scalar_t)| and |sizeof(bn254::Fr)| are same. This is
+  // because their alignment are different.
+  // See icicle/include/fields/storage.cu.h
```

Comment:
Hmm I tried but it turned out it's impossible because [the type of our `BigInt` is 64 bit](https://github.com/kroma-network/tachyon/blob/4f428ed9cebdfae448d2aa1e6ae60307ec9feb55/tachyon/math/base/big_int.h#L46-L55) and [the type of icicle `storage` is 32 bit](https://github.com/ingonyama-zk/icicle/blob/4fef5423462a72a37fe66cee89e24eae083cc112/icicle/include/fields/storage.cuh#L4-L15).